pcdsutils/import_timer.py

```diff
@@ -0,0 +1,370 @@
+        this was the main import or a python internal import used to set up
+        the import system itself. Each additional level means that this module
+        was imported by a module one level lower. For example, a module with
+        indent level 2 was imported by a module with import level 1.
+    """
+    module: str
+    root_module: str
+    self_time_raw: int
+    self_time: float
+    cumulative_time_raw: int
+    cumulative_time: float
+    indent_level: int
+
+    @classmethod
+    def from_line(cls, line: str) -> ImportTimeStats:
+        """
+        Assemble an ImportTimeStats from a python -X importtime output line.
+
+        Example lines:
+        import time:      2848 |       2848 |     pcdsdevices.stopper
+        import time:      7611 |      10459 |   pcdsdevices.valve
+        import time:      1859 |       1859 |   pcdsdevices.wfs
+        import time:      2913 |   12317383 | pcdsdevices.device_types
+        """
+        if not line.startswith('import time'):
+            raise ValueError(
+                f'Recieved {line}, which is not a valid importtime output.'
+            )
+        line = line.split('import time:')[1]
+        self_time_raw, cumulative_time_raw, module = line.split('|')
+        self_time_raw = int(self_time_raw.strip())
+        cumulative_time_raw = int(cumulative_time_raw.strip())
+        indent_level = (module.count(' ') - 1) / 2
+        module = module.strip()
+
+        return cls(
+            module=module,
+            root_module=module.split('.')[0],
+            self_time_raw=self_time_raw,
+            self_time=self_time_raw / 1e6,
+            cumulative_time_raw=cumulative_time_raw,
+            cumulative_time=cumulative_time_raw / 1e6,
+            indent_level=indent_level,
+        )
+
+
+@dataclass(frozen=True)
+class ModuleStatsSummary:
+    """
+    For a top-level module, the total time spent on all the submodules.
+
+    Attributes
+    ----------
+    root_module : str
+        The source module, e.g. pcdsdevices
+    self_time_raw : int
+        The time it took to import this module in microseconds, not counting
+        the time it took to import other modules.
+    self_time : float
+        The time it took to import this module in seconds, not counting
+        the time it took to import other modules.
+    cumulative_time_raw : int
+        The time it took to import this module in microseconds,
+        including the time it took to import other modules.
+    cumulative_time : float
+        The time it took to import this module in seconds,
+        including the time it took to import other modules.
+    submodule_stats : tuple of ImportTimeStats
+        The stats that were used to generate this stats summary.
+    """
+    root_module: str
+    self_time_raw: int
+    self_time: float
+    cumulative_time_raw: int
+    cumulative_time: float
+    submodule_stats: Tuple[ImportTimeStats]
```

Comment:
```suggestion
    submodule_stats: Tuple[ImportTimeStats, ...]
```

`List[x]` <-> `Tuple[x, ...]`

Reply:
e.g. is `Tuple[int]` a tuple of one int, while `Tuple[int, ...]` is a tuple of n ints?

Reply:
Yes, that's correct. This is because the normal usage of `Tuple` would be with (potentially) heterogeneous types in the respective positions: `Tuple[int, float, str]`